pkg/rke2/rke2.go

```diff
@@ -166,5 +175,150 @@ func setup(clx *cli.Context, cfg Config) error {
+		go startKubelet(ctx, dataDir, kubeletErr, tmpAddress, kubeletCmd)
+
+		// check for any running containers from the disabled items list
+		go checkForRunningContainers(ctx, tmpAddress, disabledItems, kubeletErr, containerdErr)
+
+		for {
+			time.Sleep(5 * time.Second)
+			select {
+			case err := <-kubeletErr:
+				logrus.Infof("kubelet Exited: %v, exiting Containerd", err)
+				// exits containerd if kubelet exits
+				containerdCmd.Process.Kill()
+				kubeletCmd.Process.Kill()
+
+			case err := <-containerdErr:
+				logrus.Infof("Containerd Exited: %v, exiting kubelet", err)
+				kubeletCmd.Process.Kill()
+				containerdCmd.Process.Kill()
+
+			case <-ctx.Done():
+				logrus.Info("Timeout reached, exiting kubelet and containerd")
+				kubeletCmd.Process.Kill()
+				containerdCmd.Process.Kill()
+			}
+			break
+		}
+	}
+
 	return nil
 }
+
+func startKubelet(ctx context.Context, dataDir string, errChan chan error, tmpAddress string, cmd *exec.Cmd) {
+	logrus.Infof("starting kubelet to clean up old static pods")
+	args := []string{
+		"--fail-swap-on=false",
+		"--container-runtime=remote",
+		"--containerd=" + tmpAddress,
+		"--container-runtime-endpoint=unix://" + tmpAddress,
+		"--pod-manifest-path=" + podManifestsDir(dataDir),
+	}
+	cmd.Args = append(cmd.Args, args...)
+	cmd.Stdout = nil
+	cmd.Stderr = nil
+
+	errChan <- cmd.Run()
+}
+
+func startContainerd(ctx context.Context, dataDir string, errChan chan error, tmpAddress string, cmd *exec.Cmd) {
+	args := []string{
+		"-a", tmpAddress,
+		"--root", filepath.Join(dataDir, "agent", "containerd"),
+	}
+	cmd.Args = append(cmd.Args, args...)
+	cmd.Stdout = nil
+	cmd.Stderr = nil
+
+	errChan <- cmd.Run()
+}
+
+func isContainerRunning(ctx context.Context, name string, resp *runtimeapi.ListContainersResponse) bool {
+	for _, c := range resp.Containers {
+		if c.Metadata.Name == name {
+			return true
+		}
+	}
+	return false
+}
+
+func checkForRunningContainers(ctx context.Context, tmpAddress string, disabledItems map[string]bool, kubeletErr, containerdErr chan error) {
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	for range ticker.C {
+		conn, err := containerdk3s.CriConnection(ctx, tmpAddress)
+		if err != nil {
+			logrus.Warnf("Failed to setup cri connection: %v", err)
+			continue
+		}
+		c := runtimeapi.NewRuntimeServiceClient(conn)
+		defer conn.Close()
+		resp, err := c.ListContainers(ctx, &runtimeapi.ListContainersRequest{})
+		if err != nil {
+			logrus.Warnf("Failed to list containers: %v", err)
+			continue
+		}
+		var gc bool
+		for item, disabled := range disabledItems {
+			if disabled {
+				if isContainerRunning(ctx, item, resp) {
+					logrus.Infof("old static pod %s is running, waiting to delete", item)
+					gc = true
+					break
+				}
+			}
+		}
+		if gc {
+			continue
+		}
+		// if all disabled items containers has been removed
+		// then terminate kubelet and containerd
+		containerdErr <- fmt.Errorf("static pods deleted exiting containerd")
```

Comment:
Since you're wrapping these error channels up above, doesn't this result in a weird log message?
> kubelet Exited: static pods deleted exiting containerd, exiting Containerd